### PR TITLE
bpf documentation updates, part 1

### PR DESCRIPTION
**Pull request by borkmann**
Related: #713

### CODE REVIEW

File: Documentation/bpf.rst [hunk 1]
```diff
@@ -989,6 +1079,85 @@ currently unsupported, too.
...
+them for compiling the BPF program whenever appropriate.
+
+A full command line example with llc's ``-mcpu=probe``:
+
+::
+
+    $ clang -O2 -Wall -target bpf -emit-llvm -c xdp-example.c -o xdp-example.bc
+    $ llc xdp-example.bc -march=bpf -mcpu=probe -filetype=obj -o xdp-example.o
+
+Generally, LLVM IR generation is mostly architecture independent. There
```
Comment (joestringer) [inline Documentation/bpf.rst]: "generally" and "mostly" here serve much the same purpose, so we probably only need one of them.

File: Documentation/bpf.rst [hunk 2]
```diff
@@ -1822,6 +1995,313 @@ pointers. After that all the programs themselves are created through the BPF
...
+bpftool is the main introspection and debugging tool around BPF and developed
+and shipped along with the Linux kernel tree under ``tools/bpf/bpftool/``.
+
+The tool can dump all BPF programs and maps that are currently loaded in
+the system, or list and correlate all BPF maps used by a specific program.
+Furthermore, it allows to dump the entire map's key / value pairs, or
+lookup, update, delete individual ones as well as retrieve a key's neighbor
+key in the map. Such operations can be performed based on BPF program or
+map IDs or by specifying the location of a BPF file system pinned program
+or map. The tool additionally also offers to pin maps or programs into the
```
Comment (joestringer) [inline Documentation/bpf.rst]: Minor nit: `... also offers` *an option* `to pin maps ...`

File: Documentation/bpf.rst [hunk 3]
```diff
@@ -1822,6 +1995,313 @@ pointers. After that all the programs themselves are created through the BPF
...
+        key 20B  value 20B  max_entries 65536  memlock 7344128B
+    7: hash  flags 0x0
+        key 10B  value 16B  max_entries 8192  memlock 790528B
+    8: hash  flags 0x0
+        key 22B  value 28B  max_entries 8192  memlock 987136B
+    9: hash  flags 0x0
+        key 20B  value 8B  max_entries 512000  memlock 49352704B
+    [...]
+
+Note that for each command, bpftool also supports json based output.
```
Comment (joestringer) [inline Documentation/bpf.rst]: What's the JSON argument? (I don't think we necessarily need an example usage, but the actual argument name might be useful)

File: Documentation/bpf.rst [hunk 4]
```diff
@@ -1822,6 +1995,313 @@ pointers. After that all the programs themselves are created through the BPF
...
+  ::
+
+     # tc filter show dev cilium_host egress
+     filter protocol all pref 1 bpf chain 0
+     filter protocol all pref 1 bpf chain 0 handle 0x1 bpf_host.o:[from-netdev] \
+                         direct-action not_in_hw id 406 tag e0362f5bd9163a0a jited
+
+The program from the object file ``bpf_host.o``, section ``from-netdev`` has
+a BPF program ID of ``406`` as denoted in ``id 406``. Based on this information
+bpftool can provide some high-level meta data specific to the program:
```
Comment (joestringer) [inline Documentation/bpf.rst]: s/meta data/metadata

**Review by joestringer — APPROVED** [review]
LGTM, thanks! Minor comments.

**Comment by borkmann** [comment]
test-me-please

**Review by raybejjani — APPROVED** [review]

**Comment by borkmann** [comment]
test-me-please

**Comment by borkmann** [comment]
test-me-please

**Comment by borkmann** [comment]
test-me-please

**Pull request merged**